libres/lib/analysis/update.cpp

```diff
@@ -67,6 +66,46 @@ typedef struct {
         *node_info; /* mutable: For the serialization of one node */
 } serialize_info_type;
 
+class update_data_type {
+public:
+    update_data_type() = default;
+    update_data_type(
+        matrix_type *S_in, matrix_type *E_in, matrix_type *D_in,
+        matrix_type *R_in, matrix_type *A_in,
+        std::vector<std::pair<matrix_type *, std::shared_ptr<RowScaling>>>
+            A_with_rowscaling_in,
+        const bool_vector_type *obs_mask_in)
+        : obs_mask(obs_mask_in) {
+        S = S_in;
+        E = E_in;
+        D = D_in;
+        R = R_in;
+        A = A_in;
+        A_with_rowscaling = A_with_rowscaling_in;
+        valid = true;
```

Comment:
What is valid?

Reply:
It is the entire `update_data_type` instance which is valid - or not. The `valid` value is based on whether we have any remaining observations - if the number of observations is zero (and hence no update should be performed), we have `valid = false`. A better name would maybe be `have_observations`

Reply:
From what you write I would suggest `has_observations` and a short descriptive comment.

Reply:
valid whether is has the data for doing an update step